test/test_mpc.py

```diff
@@ -59,6 +59,9 @@ def _check(self, encrypted_tensor, reference, msg, dst=None, tolerance=None):
             logging.info("Result - Reference = %s" % (tensor - reference))
         self.assertTrue(test_passed, msg=msg)
 
+        self.assertTrue(hasattr(encrypted_tensor, "_mac"), "_mac attribute failed to propagate")
+        self.assertEqual(encrypted_tensor._mac, 0, "_mac attribute has incorrect value")
+
```

Comment:
you could add 2 simple tests which unittest the copy_, clone, and shallow_copy methods?

Reply:
I think when it testing for these three functions([test_copy_](https://github.com/MLH-Fellowship/CrypTen/blob/7d9a070d4ca03ced5772512b22be7078e21cc2ca/test/test_mpc.py#L1520), [test_clone and shallow_copy](https://github.com/MLH-Fellowship/CrypTen/blob/7d9a070d4ca03ced5772512b22be7078e21cc2ca/test/test_mpc.py#L1509)), they have this line: `self._check(encrypted_tensor1, tensor2, "copy_ failed")` at the end, which I believe will run the checks I set in the `_check` method.

Reply:
Just some background info:

I noted in the task that these should be here as we want to make sure that adding this functionality to `copy_`, `clone` and `shallow_copy` is sufficient for all other functions to gain this functionality.

I.e. when we call something like `x.add(y)`, the code typically implements this by calling `x.clone().add_(y)`, where `add_()` is an in-place operation. My hypothesis was that because we implement most out-of-place functions by using `clone()` + in-place function, copying the `_mac` attribute in `clone`, etc. would be sufficient.

However, I wanted this unit test to ensure that we weren't missing some functions that may be implemented differently. Because all of our existing unit tests use `_check` to make the proper assertions on the output, adding these assertions to the `_check` function will propagate them to all unit tests. Our existing unit tests should sufficiently cover all functions we care about.